include/cpp2util.h

```diff
@@ -880,7 +880,7 @@ auto as( X const& x ) -> auto
     //  Experiment: Recognize the nested `::value_type` pattern for some dynamic library types
     //  like std::optional, and try to prevent accidental narrowing conversions even when
     //  those types themselves don't defend against them
-    if constexpr( requires{ typename C::value_type; } && std::is_convertible_v<X, typename C::value_type> ) {
+    if constexpr( requires{ typename C::value_type; } && requires { std::is_convertible_v<X, typename C::value_type>; } ) {
```

Comment:
Thanks! I do see the need for that `requires`.

But doesn't it also still need the original `&& std::is_convertible_c<X, typename C::value_type>` test itself?

Reply:
I have some doubts about it. When I am adding it 
```cpp
    if constexpr( requires{ typename C::value_type; } && requires { std::is_convertible_v<X, typename C::value_type>; } && std::is_convertible_v<X, typename C::value_type> ) {
```
it fails again:
```
In file included from ../tests/bug_value_type_check.cpp:2:
include/cpp2util.h:883:158: error: type 'int' cannot be used prior to '::' because it has no members
    if constexpr( requires{ typename C::value_type; } && requires { std::is_convertible_v<X, typename C::value_type>; } && std::is_convertible_v<X, typename C::value_type> ) {
                                                                                                                                                             ^
include/cpp2util.h:1474:52: note: in instantiation of function template specialization 'cpp2::as<int, char>' requested here
    else if constexpr( std::is_same_v< CPP2_TYPEOF(as<C>(CPP2_FORWARD(x))), nonesuch_ > ) {
                                                   ^
../tests/bug_value_type_check.cpp2:3:19: note: in instantiation of function template specialization 'cpp2::as_<int, char &>' requested here
    auto u {cpp2::as_<int>(x)}; 
                  ^
1 error generated.
```

Reply:
I have run all the regression tests with proposed change and they pass.

Reply:
You need to have a single _requires-expression_ with multiple *requirement*s, so
```C++
requires{
  typename C::value_type;
  requires std::is_convertible_v<X, typename C::value_type>;
}
```

Reply:
It looks like it is working well as in the next line there is another use of `C::value_type` that is not causing trouble:
```cpp
        if constexpr( is_narrowing_v<typename C::value_type, X>) {
```

So, the proposed change is passing the right `C` types through.

Reply:
> You need to have a single _requires-expression_ with multiple _requirement_s, so
> 
> ```c++
> requires{
>   typename C::value_type;
>   requires std::is_convertible_v<X, typename C::value_type>;
> }
> ```

In fact, the first _requirement_ is redundant, so `requires { requires std::is_convertible_v<X, typename C::value_type>; }` should suffice. See https://compiler-explorer.com/z/cYoETj4K3.

```C++
#include <type_traits>
struct X {
  using value_type = int;
  operator int();
};
template<class T> constexpr bool f() {
  return requires { requires std::is_convertible_v<T, typename T::value_type>; };
}
static_assert(not f<int>());
static_assert(f<X>());
```

Reply:
That is true. Pushing this version.

Reply:
Thanks @JohelEGP !

Reply:
Thanks @JohelEGP and @filipsajdak for debugging this incantation!